docs/dasom-indicator.1
.\" Manpage for dasom-indicator.
.\" Contact Hodong Kim <hodong@cogno.org>.
.TH dasom-indicator 1 "January 2016" "1.3" "dasom-indicator man page"
.SH NAME
dasom-indicator \- Input Method Framework
.SH DESCRIPTION
Dasom is an input method framework, which provides
.PP
•Input Method Server
.TP
.PP
◦dasom-indicator including XIM
.PP
•Indicators
.TP
.PP
◦GNOME shell: dasom agent extension
.TP
.PP
◦KDE, Unity, GNOME panel: dasom-indicator (based on libappindicator3)
.SH SYNOPSIS
Usage:
.TP
.PP
dasom-indicator [OPTION...] - indicator for Dasom
.SH OPTIONS
Help Options:
.TP
.PP
-h, --help
Show help options
.PP
Application Options:
.TP
.PP
--no-indicator
Do not indicatorize
.TP
.PP
--debug
Log debugging message
.TP
.PP
--version        
Version
.PP
.SH AUTHOR
Maintainer: Hodong Kim <hodong@cogno.org>


NEWS
dasom (1.2) stable; urgency=medium

  * Modified icons
  * Fixed possible faults
  * Fixed dependency on gtk3 header file (closes #33)
  * Added information: Dasom was named by Suhyeok Gwon (closes #32)
  * Removed client modules gtk, qt
    - Visit https://github.com/dasom-im/dasom-gtk for GTK+2, GTK+3 IM modules
    - Visit https://github.com/dasom-im/dasom-qt for Qt4, Qt5 IM modules

 -- Hodong Kim <hodong@cogno.org>  Mon, 30 Nov 2015 00:34:22 +0900

dasom (1.1) stable; urgency=medium

  * Fixed a possible segmentation fault that occurs when we use the inkscape
    and avoids the eclipse bug (closes #29)
  * Reduced data communications traffic (closes #4)
  * Removed Korean engine (dasom-jeongeum)
  * Moved dasom-english to libdasom as embedded engine
  * Commit and then update preedit (closes #4)
  * Do not use Delete key in deleting Jaso (closes #23)
  * Fixed race condition (closes #18)
  * Show version (closes #21)
  * Hook event only when im has focus (closes #27 #28)

 -- Hodong Kim <hodong@cogno.org>  Sun, 22 Nov 2015 00:23:11 +0900

dasom (1.0.1) stable; urgency=medium

  * Fixed race condition (closes #18)
  * Dubeolsik: Compose double consonants only with shift key by default
  * Fixed CapsLock (closes #15 #16)
  * Fixed event hooking (closes #9)
  * Fixed compatibility issue with Qt 5.5 (closes #11)
  * Modified icon brightness
    - f9f9f9 (2.5% gray) --> f2f2f2 ( 5% gray)
    - ffffff (black)     --> 1a1a1a (90% gray)

 -- Hodong Kim <hodong@cogno.org>  Sun, 25 Oct 2015 00:37:29 +0900

dasom (1.0) stable; urgency=medium

  * Initial Release

  * Dasom is an input method framework, which provides
    - Input Method Server including XIM
    - Language Engines for Korean, English
    - IM Modules for GTK+2, GTK+3, Qt4, Qt5
    - Indicators for GNOME Shell, Unity, KDE, GNOME panel

 -- Hodong Kim <hodong@cogno.org>  Fri, 09 Oct 2015 01:21:10 +0900


agents/app-indicator/dasom-indicator.c
/* -*- Mode: C; indent-tabs-mode: nil; c-basic-offset: 2; tab-width: 2 -*-  */
/*
 * dasom-indicator.c
 * This file is part of Dasom.
 *
 * Copyright (C) 2015,2016 Hodong Kim <cogniti@gmail.com>
 *
 * Dasom is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Dasom is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program;  If not, see <http://www.gnu.org/licenses/>.
 */

#include "config.h"
#include <glib/gi18n.h>
#include <gtk/gtk.h>
#include "dasom.h"
#include <gio/gunixsocketaddress.h>
#include <libappindicator/app-indicator.h>
#include <syslog.h>
#include "dasom-private.h"
#include <stdlib.h>

gboolean syslog_initialized = FALSE;

static void on_about_menu (GtkWidget *widget,
                           gpointer   data)
{
  g_debug (G_STRLOC ": %s", G_STRFUNC);

  gchar *artists[]     = {"Hodong Kim <cogniti@gmail.com>", NULL};
  gchar *authors[]     = {"Hodong Kim <cogniti@gmail.com>", NULL};
  gchar *documenters[] = {"Hodong Kim <cogniti@gmail.com>", NULL};

  gtk_show_about_dialog (NULL,
    "artists",            artists,
    "authors",            authors,
    "comments",           _("An indicator for Dasom\nDasom was named by Suhyeok Gwon."),
    "copyright",          "Copyright (c) 2015,2016 Hodong Kim",
    "documenters",        documenters,
    "license-type",       GTK_LICENSE_LGPL_3_0,
    "logo-icon-name",     "dasom-indicator",
    "program-name",       _("Dasom Indicator"),
    "translator-credits", "Hodong Kim <cogniti@gmail.com>",
    "version",            VERSION,
    "website",            "https://dasom-im.github.io",
    "website-label",      _("Website"),
    NULL);
}

static void on_exit_menu (GtkWidget *widget,
                          gpointer   data)
{
  g_debug (G_STRLOC ": %s", G_STRFUNC);

  gtk_main_quit ();
}

/* FIXME: need DasomEngineInfo or id instead of name */
static void on_engine_changed (DasomAgent   *agent,
                               gchar        *name,
                               AppIndicator *indicator)
{
  g_debug (G_STRLOC ": %s: name: %s", G_STRFUNC, name);

  gchar *icon_name;

  if (g_strcmp0 (name, "focus-out") == 0)
    icon_name = g_strdup ("dasom-indicator");
  else
    icon_name = g_strdup_printf ("dasom-indicator-%s", name);

  app_indicator_set_icon_full (indicator, icon_name, name);

  g_free (icon_name);
}

static void on_disconnected (DasomAgent   *agent,
                             AppIndicator *indicator)
{
  g_debug (G_STRLOC ": %s", G_STRFUNC);

  app_indicator_set_icon_full (indicator,
                               "dasom-indicator-warning", "disconnected");
}

int
main (int argc, char **argv)
{
  GError         *error     = NULL;
  gboolean        is_no_daemon = FALSE;
  gboolean        is_debug     = FALSE;
  gboolean        is_version   = FALSE;
  GOptionContext *context;
  GOptionEntry    entries[] = {
    {"no-daemon", 0, 0, G_OPTION_ARG_NONE, &is_no_daemon, N_("Do not daemonize"), NULL},
    {"debug", 0, 0, G_OPTION_ARG_NONE, &is_debug, N_("Log debugging message"), NULL},
    {"version", 0, 0, G_OPTION_ARG_NONE, &is_version, N_("Version"), NULL},
    {NULL}
  };

  context = g_option_context_new ("- indicator for Dasom");
  g_option_context_add_main_entries (context, entries, GETTEXT_PACKAGE);
  g_option_context_parse (context, &argc, &argv, &error);
  g_option_context_free (context);

  if (error != NULL)
  {
    g_warning ("%s", error->message);
    g_error_free (error);
    return EXIT_FAILURE;
  }

#ifdef ENABLE_NLS
  bindtextdomain (GETTEXT_PACKAGE, DASOM_LOCALE_DIR);
  bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
  textdomain (GETTEXT_PACKAGE);
#endif

  if (is_debug)
    g_setenv ("G_MESSAGES_DEBUG", "dasom", TRUE);

  if (is_version)
  {
    g_print ("%s %s\n", argv[0], VERSION);
    exit (EXIT_SUCCESS);
  }

  if (is_no_daemon == FALSE)
  {
    openlog (g_get_prgname (), LOG_PID | LOG_PERROR, LOG_DAEMON);
    syslog_initialized = TRUE;
    g_log_set_default_handler ((GLogFunc) dasom_log_default_handler, &is_debug);

    if (daemon (0, 0) != 0)
    {
      g_critical ("Couldn't daemonize.");
      return EXIT_FAILURE;
    }
  }

  gtk_init (&argc, &argv);

  AppIndicator *indicator;
  GtkWidget    *menu_shell;
  GtkWidget    *about_menu;
  GtkWidget    *exit_menu;
  DasomAgent   *agent;

  menu_shell = gtk_menu_new ();
  about_menu = gtk_menu_item_new_with_label (_("About"));
  exit_menu  = gtk_menu_item_new_with_label (_("Exit"));

  gtk_widget_show_all (about_menu);
  gtk_widget_show_all (exit_menu);
  gtk_menu_shell_append (GTK_MENU_SHELL (menu_shell), about_menu);
  gtk_menu_shell_append (GTK_MENU_SHELL (menu_shell), exit_menu);

  gtk_widget_show_all (menu_shell);

  indicator = app_indicator_new ("dasom-indicator", "input-keyboard",
                                 APP_INDICATOR_CATEGORY_APPLICATION_STATUS);
  app_indicator_set_status (indicator, APP_INDICATOR_STATUS_ACTIVE);
  app_indicator_set_icon_full (indicator, "dasom-indicator", "Dasom");
  app_indicator_set_menu (indicator, GTK_MENU (menu_shell));

  g_signal_connect (about_menu, "activate",
                    G_CALLBACK (on_about_menu), indicator);
  g_signal_connect (exit_menu,  "activate",
                    G_CALLBACK (on_exit_menu),  indicator);

  agent = dasom_agent_new ();

  g_signal_connect (agent, "engine-changed",
                    G_CALLBACK (on_engine_changed), indicator);
  g_signal_connect (agent, "disconnected",
                    G_CALLBACK (on_disconnected), indicator);

  dasom_agent_connect_to_server (agent);

  gtk_main ();

  g_object_unref (agent);
  g_object_unref (indicator);

  if (syslog_initialized)
    closelog ();

  return EXIT_SUCCESS;
}
